darwin/torch/dataset.py

```diff
@@ -0,0 +1,276 @@
+    elif mode == "instance_segmentation":
+        import darwin.torch.transforms as T
+
+        trfs = [T.ConvertPolysToInstanceMasks()]
+        if transforms is not None:
+            trfs.append(transforms)
+        transforms = T.Compose(trfs)
+        dataset = InstanceSegmentationDataset(
+            root, image_set=image_set, split_id=split_id, transforms=transforms
+        )
+    elif mode == "semantic_segmentation":
+        import darwin.torch.transforms as T
+
+        trfs = [T.ConvertPolysToMask()]
+        if transforms is not None:
+            trfs.append(transforms)
+        transforms = T.Compose(trfs)
+        dataset = SemanticSegmentationDataset(
+            root, image_set=image_set, split_id=split_id, transforms=transforms
+        )
+    else:
+        raise ValueError("Dataset type {mode} not supported.")
+
+    return dataset
+
+
+class Dataset(object):
+    def __init__(self, root: Path, image_set: str, split_id: Optional[str] = None, transforms=None):
+        self.root = root
+        self.transforms = transforms
+        self.image_set = image_set
+
+        if self.image_set not in ["train", "val", "test"]:
+            raise ValueError("Unknown partition {self.image_set}")
+
+        path_to_lists = root / "lists"
+        if split_id is not None:
+            path_to_lists /= split_id
+        file_partition = path_to_lists / f"{image_set}.txt"
+
+        if not file_partition.exists():
+            raise FileNotFoundError(
+                "Could not find partition {image_set} in {path_to_lists}. (Is the percentage larger than 0?)"
+            )
+        stems = [e.strip() for e in open(file_partition)]
+
+        exts = ["jpg", "jpeg", "png"]
+        self.annotations, self.images = [], []
+        for s in stems:
+            annot_path = root / f"annotations/{s}.json"
+            for ext in exts:
+                im_path = root / f"images/{s}.{ext}"
+                if im_path.is_file():
+                    self.images.append(im_path)
+                    self.annotations.append(annot_path)
+                    break
+        if len(self.images) == 0:
+            raise ValueError(f"could not find any {exts} file in {root/'images'}")
+
+    def __getitem__(self, idx: int):
+        # load images ad masks
+        img = load_pil_image(self.images[idx])
+        target = self._load_anno_and_remap(idx)
+
+        if self.transforms is not None:
+            img, target = self.transforms(img, target)
+
+        return img, target
+
+    def _load_anno_and_remap(self, idx: int):
+        with open(self.annotations[idx]) as f:
+            anno = json.load(f)["annotations"]
+
+        # Filter out unused classes
+        anno = [obj for obj in anno if obj["name"] in self.classes]
+
+        res = dict(image_id=idx, annotations=anno)
+        return res
+
+    def __len__(self):
+        return len(self.images)
+
+    def subsample(self, perc: float):
+        num = int(len(self.annotations) * perc)
```

Comment:
this is not rounding but is flooring the float, is this intentional? If all split use floor there is a chance that some images will be missed